File: src/main/java/tool/cmnclslib/cls/ClsConfigFile.java
package tool.cmnclslib.cls;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import tool.cmnclslib.ifc.ICmnLogger;
import tool.cmnclslib.mdl.MdlConst;
import tool.cmnclslib.mdl.MdlFile;
import tool.cmnclslib.mdl.MdlUtil;

/**
 * 設定ファイル（Key=Value形式および行リスト形式）を読み込み、解析・保持するためのクラスです。
 */
public class ClsConfigFile {

    private static final Pattern COMMENT_REGEX = Pattern.compile("^\\s*#");
    private static final Pattern EMPTY_LINE_REGEX = Pattern.compile("^\\s*$");
    private static final Pattern INLINE_COMMENT_REGEX = Pattern.compile("#.+");

    private ICmnLogger logger;
    private Map<String, String> configMap = new LinkedHashMap<>();
    private Map<String, List<String>> listMap = new LinkedHashMap<>();
    private List<String> duplicateKeys = new ArrayList<>();
    private List<String> configList = new ArrayList<>();
    private int verbose = 0;
    private String pattern = "^(?<KEY>[^#=]+)=(?<VALUE>.+)$";
    private Pattern cachedPatternRegex;
    private Charset encoding = Charset.defaultCharset();
    private boolean isSkipComment = true;
    private boolean detectEncoding = true;

    /**
     * ClsConfigFile クラスの新しいインスタンスを初期化します。
     *
     * @param logger ログ出力に使用するロガーオブジェクト
     */
    public ClsConfigFile(ICmnLogger logger) {
        this.logger = logger;
        clear();
    }

    public Map<String, String> getConfigMap() {
        return configMap;
    }

    public void setConfigMap(Map<String, String> configMap) {
        this.configMap = configMap != null ? configMap : new LinkedHashMap<>();
    }

    /**
     * @deprecated {@link #getConfigMap()} を使用してください。
     */
    @Deprecated
    public Map<String, String> getConfigDictionary() {
        return getConfigMap();
    }

    /**
     * @deprecated {@link #setConfigMap(Map)} を使用してください。
     */
    @Deprecated
    public void setConfigDictionary(Map<String, String> configDictionary) {
        setConfigMap(configDictionary);
    }

    public Map<String, List<String>> getListMap() {
        return listMap;
    }

    public void setListMap(Map<String, List<String>> listMap) {
        this.listMap = listMap != null ? listMap : new LinkedHashMap<>();
    }

    /**
     * @deprecated {@link #getListMap()} を使用してください。
     */
    @Deprecated
    public Map<String, List<String>> getListDictionary() {
        return getListMap();
    }

    /**
     * @deprecated {@link #setListMap(Map)} を使用してください。
     */
    @Deprecated
    public void setListDictionary(Map<String, List<String>> listDictionary) {
        setListMap(listDictionary);
    }

    public List<String> getConfigList() {
        return configList;
    }

    public void setConfigList(List<String> configList) {
        this.configList = configList != null ? configList : new ArrayList<>();
    }

    public List<String> getDuplicateKeys() {
        return duplicateKeys;
    }

    public void setDuplicateKeys(List<String> duplicateKeys) {
        this.duplicateKeys = duplicateKeys != null ? duplicateKeys : new ArrayList<>();
    }

    public int getVerbose() {
        return verbose;
    }

    public void setVerbose(int verbose) {
        this.verbose = verbose;
    }

    public String getPattern() {
        return pattern;
    }

    public void setPattern(String pattern) {
        this.pattern = pattern;
        this.cachedPatternRegex = null;
    }

    public Charset getEncoding() {
        return encoding;
    }

    public void setEncoding(Charset encoding) {
        this.encoding = encoding;
    }

    public boolean isSkipComment() {
        return isSkipComment;
    }

    public void setSkipComment(boolean skipComment) {
        this.isSkipComment = skipComment;
    }

    public boolean isDetectEncoding() {
        return detectEncoding;
    }

    public void setDetectEncoding(boolean detectEncoding) {
        this.detectEncoding = detectEncoding;
    }

    /**
     * 保持している設定情報（マップ、リスト等）をすべてクリアします。
     */
    public void clear() {
        configMap.clear();
        listMap.clear();
        configList.clear();
        duplicateKeys.clear();
    }

    /**
     * 指定された設定ファイルを読み込み、設定内容を Key-Value のマップに格納します。
     *
     * @param filePath 読み込み対象の設定ファイルパス
     * @return 正常に読み込んでマップに格納された設定項目数。エラー発生時は -1
     */
    public int loadToMap(String filePath) {
        final String methodName = "[ClsConfigFile.LoadToMap()]";
        if (verbose > 3) {
            writeLog(MdlConst.LVL_DEBUG, methodName);
            writeLog(MdlConst.LVL_DEBUG, methodName + "filePath       = " + filePath);
            writeLog(MdlConst.LVL_DEBUG, methodName + "verbose       = " + verbose);
            writeLog(MdlConst.LVL_DEBUG, methodName + "isSkipComment = " + isSkipComment);
            writeLog(MdlConst.LVL_DEBUG, methodName);
        }

        if (configMap == null) {
            writeLog(MdlConst.LVL_E, methodName + " configMap is null");
            return -1;
        }

        try {
            if (detectEncoding) {
                encoding = MdlFile.detectFileEncoding(filePath);
            }

            if (cachedPatternRegex == null) {
                cachedPatternRegex = Pattern.compile(pattern);
            }

            java.nio.file.Path path = java.nio.file.Paths.get(filePath);
            Charset enc = encoding != null ? encoding : Charset.defaultCharset();
            try (BufferedReader reader = java.nio.file.Files.newBufferedReader(path, enc)) {
                String line;
                while ((line = reader.readLine()) != null) {
                    String buffer;
                    String lineType = "NORMAL LINE";
                    boolean isContinue = false;

                    // コメント行
                    if (isSkipComment) {
                        if (COMMENT_REGEX.matcher(line).find()) {
                            isContinue = true;
                            lineType = "SKIP : COMMENT LINE";
                        }
                        buffer = INLINE_COMMENT_REGEX.matcher(line).replaceAll("").trim();
                    } else {
                        buffer = line.trim();
                    }

                    // 空行
                    if (EMPTY_LINE_REGEX.matcher(buffer).find()) {
                        isContinue = true;
                        lineType = "SKIP : EMPTY LINE";
                    }

                    if (verbose > 5) {
                        writeLog(MdlConst.LVL_DEBUG, methodName + "CURRENT LINE (" + lineType + ") : " + buffer);
                    }

                    if (isContinue) {
                        continue;
                    }

                    Matcher match = cachedPatternRegex.matcher(buffer);
                    if (match.find()) {
                        String key = MdlUtil.trimQuotes(match.group("KEY"));
                        String value = MdlUtil.trimQuotes(match.group("VALUE"));

                        if (verbose > 3) {
                            writeLog(MdlConst.LVL_DEBUG, methodName + "[" + filePath + "] configMap[" + key + "] = " + value);
                        }

                        if (!duplicateKeys.isEmpty() && duplicateKeys.contains(key)) {
                            listMap.computeIfAbsent(key, k -> new ArrayList<>()).add(value);
                        }

                        configMap.put(key, value);
                    }
                }
            }
            return configMap.size();
        } catch (Exception ex) {
            writeLog(MdlConst.LVL_E, methodName + " Exception : " + ex.getMessage());
            return -1;
        }
    }

    /**
     * @deprecated {@link #loadToMap(String)} を使用してください。
     */
    @Deprecated
    public int loadToDictionary(String filePath) {
        return loadToMap(filePath);
    }

    /**
     * 指定された設定ファイルを読み込み、各行の文字列をリストに格納します。
     *
     * @param filePath 読み込み対象の設定ファイルパス
     * @param unique 重複行を除外してユニークな行のみ保持する場合は true
     * @return 正常に読み込んでリストに格納された行数。エラー発生時は -1
     */
    public int loadToList(String filePath, boolean unique) {
        final String methodName = "[ClsConfigFile.LoadToList()]";
        if (verbose > 3) {
            writeLog(MdlConst.LVL_DEBUG, methodName);
            writeLog(MdlConst.LVL_DEBUG, methodName + "filePath              = " + filePath);
            writeLog(MdlConst.LVL_DEBUG, methodName + "verbose              = " + verbose);
            writeLog(MdlConst.LVL_DEBUG, methodName + "isSkipComment = " + isSkipComment);
            writeLog(MdlConst.LVL_DEBUG, methodName);
        }

        if (configList == null) {
            writeLog(MdlConst.LVL_E, methodName + " configList is null");
            return -1;
        }

        try {
            if (detectEncoding) {
                encoding = MdlFile.detectFileEncoding(filePath);
            }

            java.nio.file.Path path = java.nio.file.Paths.get(filePath);
            Charset enc = encoding != null ? encoding : Charset.defaultCharset();
            try (BufferedReader reader = java.nio.file.Files.newBufferedReader(path, enc)) {
                String line;
                while ((line = reader.readLine()) != null) {
                    String buffer;
                    String lineType = "NORMAL LINE";
                    boolean isContinue = false;

                    if (isSkipComment) {
                        if (COMMENT_REGEX.matcher(line).find()) {
                            isContinue = true;
                            lineType = "SKIP : COMMENT LINE";
                        }
                        buffer = INLINE_COMMENT_REGEX.matcher(line).replaceAll("").trim();
                    } else {
                        buffer = line.trim();
                    }

                    if (EMPTY_LINE_REGEX.matcher(buffer).find()) {
                        isContinue = true;
                        lineType = "SKIP : EMPTY LINE";
                    }

                    if (verbose > 5) {
                        writeLog(MdlConst.LVL_DEBUG, methodName + "CURRENT LINE (" + lineType + ") : " + buffer);
                    }

                    if (isContinue) {
                        continue;
                    }

                    if (!unique || !configList.contains(buffer)) {
                        configList.add(buffer);
                    }
                }
            }
            return configList.size();
        } catch (Exception ex) {
            writeLog(MdlConst.LVL_E, methodName + " Exception : " + ex.getMessage());
            return -1;
        }
    }

    /**
     * 指定されたレベルとメッセージでログを出力します。
     *
     * @param level ログレベル
     * @param message 出力するメッセージ文字列
     */
    public void writeLog(int level, String message) {
        if (logger != null) {
            logger.writeLine(level, message);
        } else {
            System.out.println(message);
        }
    }
}

File: CmnClsLib/Class/ClsConfigFile.cs
using System;
using System.Text;
using System.Text.RegularExpressions;
using CmnClsLib.Module;
using CmnClsLib.Interface;

// 2026/08/08 Gemini 3.6 Flash (High) Review & Modified

namespace CmnClsLib.Class
{
    /// <summary>
    /// 設定ファイル（Key=Value形式および行リスト形式）を読み込み、解析・保持するためのクラスです。
    /// </summary>
    /// <example>
    /// <code>
    /// var logger = new CmnLogger();
    /// var configFile = new ClsConfigFile(logger);
    /// int loadedCount = configFile.LoadToDictionary(@"C:\config.ini");
    /// string value = configFile.ConfigDictionary["SettingKey"];
    /// </code>
    /// </example>
    public partial class ClsConfigFile
    {
        [GeneratedRegex(@"^\s*#")]
        private static partial Regex GeneratedRegexComment();

        [GeneratedRegex(@"^\s*$")]
        private static partial Regex GeneratedRegexEmpty();

        [GeneratedRegex(@"#.+")]
        private static partial Regex GeneratedRegexInlineComment();

        private ICmnLogger? _logger;
        private Dictionary<string, string> _configDictionary = [];
        private Dictionary<string, List<string>> _listDictionary = [];
        private List<string> _duplicateKeys = [];
        private List<string> _configList = [];
        private int _verbose = 0;
        private string _pattern = "^(?<KEY>[^#=]+)=(?<VALUE>.+)$";
        private Regex? _cachedPatternRegex;
        private Encoding _encoding = Encoding.Default;
        private readonly Regex _commentRegex = GeneratedRegexComment();
        private readonly Regex _emptyLineRegex = GeneratedRegexEmpty();
        private readonly Regex _inlineCommentRegex = GeneratedRegexInlineComment();
        private bool _isSkipComment = true;
        private bool _detectEncoding = true;

        static ClsConfigFile()
        {
            Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
        }

        /// <summary>
        /// <see cref="ClsConfigFile"/> クラスの新しいインスタンスを初期化します。
        /// </summary>
        /// <param name="logger">ログ出力に使用するロガーオブジェクト。</param>
        /// <example>
        /// <code>
        /// var configFile = new ClsConfigFile(logger);
        /// </code>
        /// </example>
        public ClsConfigFile(ICmnLogger logger)
        {
            _logger = logger;
            Clear();
        }

        /// <summary>
        /// 読み込んだ設定キーと値のペアを格納する辞書を取得または設定します。
        /// </summary>
        public Dictionary<string, string> ConfigDictionary { get => _configDictionary; set => _configDictionary = value; }

        /// <summary>
        /// 重複を許可した特定キーに対する複数の値を格納する辞書を取得または設定します。
        /// </summary>
        public Dictionary<string, List<string>> ListDictionary { get => _listDictionary; set => _listDictionary = value; }

        /// <summary>
        /// 読み込んだ設定行の一覧を取得または設定します。
        /// </summary>
        public List<string> ConfigList { get => _configList; set => _configList = value; }

        /// <summary>
        /// 重複値をリスト形式 (<see cref="ListDictionary"/>) で保持することを許可するキーのリストを取得または設定します。
        /// </summary>
        public List<string> DuplicateKeys { get => _duplicateKeys; set => _duplicateKeys = value; }

        /// <summary>
        /// 詳細ログの出力レベルを取得または設定します（数値が大きいほど詳細なログが出力されます）。
        /// </summary>
        public int Verbose { get => _verbose; set => _verbose = value; }

        /// <summary>
        /// Key=Value を抽出するための正規表現パターンを取得または設定します。
        /// </summary>
        public string Pattern
        {
            get => _pattern;
            set
            {
                _pattern = value;
                _cachedPatternRegex = null;
            }
        }

        /// <summary>
        /// ファイル読み込み時に使用する文字エンコーディングを取得または設定します。
        /// </summary>
        public Encoding Encoding { get => _encoding; set => _encoding = value; }

        /// <summary>
        /// コメント行（#で始まる行および行末コメント）をスキップするかどうかを取得または設定します。
        /// </summary>
        public bool IsSkipComment { get => _isSkipComment; set => _isSkipComment = value; }

        #region 旧互換用プロパティ
        /// <summary>
        /// 読み込んだ設定キーと値のペアを格納する辞書を取得または設定します。
        /// </summary>
        [Obsolete("代わりに 'ConfigDictionary' を使用します。")]
        public Dictionary<string, string> ConfigDic { get => ConfigDictionary; set => ConfigDictionary = value; }

        /// <summary>
        /// 重複を許可した特定キーに対する複数の値を格納する辞書を取得または設定します。
        /// </summary>
        [Obsolete("代わりに 'ListDictionary' を使用します。")]
        public Dictionary<string, List<string>> ListDic { get => ListDictionary; set => ListDictionary = value; }

        /// <summary>
        /// 重複値をリスト形式で保持することを許可するキーのリストを取得または設定します。
        /// </summary>
        [Obsolete("代わりに 'DuplicateKeys' を使用します。")]
        public List<string> DuplicateKeyList { get => DuplicateKeys; set => DuplicateKeys = value; }
        #endregion

        /// <summary>
        /// 保持している設定情報（辞書、リスト等）をすべてクリアします。
        /// </summary>
        /// <example>
        /// <code>
        /// configFile.Clear();
        /// </code>
        /// </example>
        public void Clear()
        {
            _configDictionary.Clear();
            _listDictionary.Clear();
            _configList.Clear();
            _duplicateKeys.Clear();
        }

        /// <summary>
        /// 指定された設定ファイルを読み込み、設定内容を Key-Value の辞書 (<see cref="ConfigDictionary"/>) に格納します。
        /// </summary>
        /// <param name="filePath">読み込み対象の設定ファイルパス。</param>
        /// <returns>正常に読み込んで辞書に格納された設定項目数。エラー発生時は -1。</returns>
        /// <example>
        /// <code>
        /// var configFile = new ClsConfigFile(logger);
        /// int count = configFile.LoadToDictionary(@"C:\app.conf");
        /// Console.WriteLine($"Loaded {count} items.");
        /// </code>
        /// </example>
        public int LoadToDictionary(string filePath)
        {
            const string METHOD_NAME = "[ClsConfigFile.LoadToDictionary()]";
            if (_verbose > 3)
            {
                WriteLog(MdlConst.LVL_DEBUG, METHOD_NAME);
                WriteLog(MdlConst.LVL_DEBUG, $"{METHOD_NAME}filePath       = {filePath}");
                WriteLog(MdlConst.LVL_DEBUG, $"{METHOD_NAME}_verbose       = {_verbose}");
                WriteLog(MdlConst.LVL_DEBUG, $"{METHOD_NAME}_isSkipComment = {_isSkipComment}");
                WriteLog(MdlConst.LVL_DEBUG, METHOD_NAME);
            }

            if (_configDictionary is null)
            {
                WriteLog(MdlConst.LVL_E, $"{METHOD_NAME} _configDictionary is null");
                return -1;
            }

            try
            {
                if (_detectEncoding)
                {
                    _encoding = MdlFile.DetectFileEncoding(filePath);
                }

                using var reader = new StreamReader(filePath, _encoding ?? Encoding.Default);
                _cachedPatternRegex ??= new Regex(_pattern, RegexOptions.Compiled);

                string? line;
                while ((line = reader.ReadLine()) is not null)
                {
                    string buffer;
                    string lineType = "NORMAL LINE";
                    bool isContinue = false;

                    // コメント行
                    if (_isSkipComment)
                    {
                        if (_commentRegex.IsMatch(line))
                        {
                            isContinue = true;
                            lineType = "SKIP : COMMENT LINE";
                        }
                        buffer = _inlineCommentRegex.Replace(line, "").Trim();
                    }
                    else
                    {
                        buffer = line.Trim();
                    }

                    // 空行
                    if (_emptyLineRegex.IsMatch(buffer))
                    {
                        isContinue = true;
                        lineType = "SKIP : EMPTY LINE";
                    }

                    if (_verbose > 5)
                    {
                        WriteLog(MdlConst.LVL_DEBUG, $"{METHOD_NAME}CURRENT LINE ({lineType}) : {buffer}");
                    }

                    if (isContinue) continue;

                    Match match = _cachedPatternRegex.Match(buffer);
                    if (match.Success)
                    {
                        string key = MdlUtil.TrimQuotes(match.Groups["KEY"].Value);
                        string value = MdlUtil.TrimQuotes(match.Groups["VALUE"].Value);

                        if (_verbose > 3)
                        {
                            WriteLog(MdlConst.LVL_DEBUG, $"{METHOD_NAME}[{filePath}] configDictionary[{key}] = {value}");
                        }

                        if (_duplicateKeys.Count > 0 && _duplicateKeys.Contains(key))
                        {
                            if (!_listDictionary.TryGetValue(key, out var tempList))
                            {
                                tempList = [];
                                _listDictionary[key] = tempList;
                            }
                            tempList.Add(value);
                        }

                        _configDictionary[key] = value;
                    }
                }
                return _configDictionary.Count;
            }
            catch (Exception ex)
            {
                WriteLog(MdlConst.LVL_E, $"{METHOD_NAME} Exception : {ex.Message}");
                return -1;
            }
        }

        /// <summary>
        /// 設定ファイルを読み込み、辞書に格納します（旧形式互換用メソッド）。
        /// </summary>
        /// <param name="filePath">設定ファイルのパス。</param>
        /// <returns>読み込んだ行数。</returns>
        /// <example>
        /// <code>
        /// int count = configFile.ReadFile(@"C:\app.conf");
        /// </code>
        /// </example>
        [Obsolete("代わりに 'LoadToDictionary()' を使用します。")]
        public int ReadFile(string filePath) => LoadToDictionary(filePath);

        /// <summary>
        /// 指定された設定ファイルを読み込み、各行の文字列をリスト (<see cref="ConfigList"/>) に格納します。
        /// </summary>
        /// <param name="filePath">読み込み対象の設定ファイルパス。</param>
        /// <param name="unique">重複行を除外してユニークな行のみ保持する場合は <c>true</c>。</param>
        /// <returns>正常に読み込んでリストに格納された行数。エラー発生時は -1。</returns>
        /// <example>
        /// <code>
        /// var configFile = new ClsConfigFile(logger);
        /// int lineCount = configFile.LoadToList(@"C:\list.txt", unique: true);
        /// </code>
        /// </example>
        public int LoadToList(string filePath, bool unique)
        {
            const string METHOD_NAME = "[ClsConfigFile.LoadToList()]";
            if (_verbose > 3)
            {
                WriteLog(MdlConst.LVL_DEBUG, METHOD_NAME);
                WriteLog(MdlConst.LVL_DEBUG, $"{METHOD_NAME}filePath              = {filePath}");
                WriteLog(MdlConst.LVL_DEBUG, $"{METHOD_NAME}_verbose              = {_verbose}");
                WriteLog(MdlConst.LVL_DEBUG, $"{METHOD_NAME}_isSkipComment = {_isSkipComment}");
                WriteLog(MdlConst.LVL_DEBUG, METHOD_NAME);
            }

            if (_configList is null)
            {
                WriteLog(MdlConst.LVL_E, $"{METHOD_NAME} _configList is null");
                return -1;
            }

            try
            {
                if (_detectEncoding)
                {
                    _encoding = MdlFile.DetectFileEncoding(filePath);
                }

                using var reader = new StreamReader(filePath, _encoding ?? Encoding.Default);

                string? line;
                while ((line = reader.ReadLine()) is not null)
                {
                    string buffer;
                    string lineType = "NORMAL LINE";
                    bool isContinue = false;

                    // コメント
                    if (_isSkipComment)
                    {
                        if (_commentRegex.IsMatch(line))
                        {
                            isContinue = true;
                            lineType = "SKIP : COMMENT LINE";
                        }
                        buffer = _inlineCommentRegex.Replace(line, "").Trim();
                    }
                    else
                    {
                        buffer = line.Trim();
                    }

                    // 空行
                    if (_emptyLineRegex.IsMatch(buffer))
                    {
                        isContinue = true;
                        lineType = "SKIP : EMPTY LINE";
                    }

                    if (_verbose > 5)
                    {
                        WriteLog(MdlConst.LVL_DEBUG, $"{METHOD_NAME}CURRENT LINE ({lineType}) : {buffer}");
                    }

                    if (isContinue) continue;

                    if (!unique || !_configList.Contains(buffer))
                    {
                        _configList.Add(buffer);
                    }
                }
                return _configList.Count;
            }
            catch (Exception ex)
            {
                WriteLog(MdlConst.LVL_E, $"{METHOD_NAME} Exception : {ex.Message}");
                return -1;
            }
        }

        /// <summary>
        /// 設定ファイルを読み込み、リストに格納します（旧形式互換用メソッド）。
        /// </summary>
        /// <param name="filePath">設定ファイルのパス。</param>
        /// <param name="unique">重複行を除外するかどうか。</param>
        /// <returns>読み込んだ行数。</returns>
        /// <example>
        /// <code>
        /// int count = configFile.ReadFileToList(@"C:\list.txt", true);
        /// </code>
        /// </example>
        [Obsolete("代わりに 'LoadToList()' を使用します。")]
        public int ReadFileToList(string filePath, bool unique) => LoadToList(filePath, unique);

        /// <summary>
        /// 指定されたレベルとメッセージでログを出力します。ロガー未設定の場合は標準出力へ出力します。
        /// </summary>
        /// <param name="level">ログレベル。</param>
        /// <param name="message">出力するメッセージ文字列。</param>
        /// <example>
        /// <code>
        /// configFile.WriteLog(MdlConst.LVL_DEBUG, "処理を開始しました。");
        /// </code>
        /// </example>
        public void WriteLog(int level, string message)
        {
            if (_logger != null)
            {
                _logger.WriteLine(level, message);
            }
            else
            {
                try
                {
                    Console.WriteLine(message);
                }
                catch { }
            }
        }

        /// <summary>
        /// ログメッセージを書き込みます（旧形式互換用メソッド）。
        /// </summary>
        /// <param name="level">エラーレベル。</param>
        /// <param name="message">ログメッセージ。</param>
        /// <example>
        /// <code>
        /// configFile.Writeln(1, "メッセージ");
        /// </code>
        /// </example>
        [Obsolete("代わりに 'WriteLog()' を使用します。")]
        public void Writeln(int level, string message) => WriteLog(level, message);
    }
}
